pkg/lib/serverTime.js

```diff
@@ -321,13 +301,119 @@ export function ServerTime() {
                     return result;
                 });
-    };
+    }
+
+    function set_custom_ntp_timesyncd(servers, enabled) {
+        const custom_ntp_config_file = cockpit.file("/etc/systemd/timesyncd.conf.d/50-cockpit.conf",
+                                                    { superuser: "try" });
 
-    self.set_custom_ntp = function (servers, enabled) {
         const text = `# This file is automatically generated by Cockpit\n\n[Time]\n${enabled ? "" : "#"}NTP=${servers.join(" ")}\n`;
 
         return cockpit.spawn(["mkdir", "-p", "/etc/systemd/timesyncd.conf.d"], { superuser: "try" })
                 .then(() => custom_ntp_config_file.replace(text));
+    }
+
+    function get_custom_ntp_chronyd() {
+        const result = {
+            supported: true,
+        };
+
+        const enabled_file = cockpit.file("/etc/chrony-cockpit-sources/custom.sources", { superuser: "try" });
+        const disabled_file = cockpit.file("/etc/chrony-cockpit-sources/custom.disabled", { superuser: "try" });
+
+        function parse_servers(data) {
+            console.log("P", data);
+            const servers = [];
+            data.split("\n").forEach(function(line) {
+                const parts = line.split(" ");
+                if (parts[0] == "server")
+                    servers.push(parts[1]);
+            });
+            return servers;
+        }
+
+        return enabled_file.read()
+                .then(data => {
+                    if (data) {
+                        result.servers = parse_servers(data);
+                        result.enabled = true;
+                        return Promise.resolve(result);
+                    } else {
+                        return disabled_file.read()
+                                .then(data => {
+                                    result.servers = parse_servers(data);
+                                    result.enabled = false;
+                                    return Promise.resolve(result);
+                                });
+                    }
+                });
+    }
+
+    function set_custom_ntp_chronyd(servers, enabled) {
+        const enabled_file = cockpit.file("/etc/chrony-cockpit-sources/custom.sources", { superuser: "try" });
+        const disabled_file = cockpit.file("/etc/chrony-cockpit-sources/custom.disabled", { superuser: "try" });
+
+        const text = "# This file is automatically generated by Cockpit\n\n" + servers.map(s => `server ${s}\n`).join("");
+
+        function ensure_sourcedir() {
+            function add_sourcedir(data) {
+                console.log("D", data);
+                if (data.indexOf("sourcedir /etc/chrony-cockpit-sources") == -1)
+                    data += "\n# Sources added by the Cockpit web console\nsourcedir /etc/chrony-cockpit-sources\n";
+                return data;
+            }
+            return cockpit.file("/etc/chrony.conf", { superuser: "try" }).modify(add_sourcedir);
```

Comment:
Without testing this, I found the man page of [chrony](https://man.archlinux.org/man/community/chrony/chronyd.8.en#OPTIONS) mentions a `-p` option which prints its config and is supposed to be used for verifying if it's correct. So maybe something we should add, or write to a temporary config file => verify => write final file.

Reply:
Hmm, that would only be useful to detect bugs in our own code, no?

Reply:
But what if the user has edited chrony.conf? We could end up with a broken config file then right?

Reply:
Sure, but it was broken before as well, no?

Reply:
Yes, then we will keep it broken. I guess as we add a new file that's ok?

Reply:
I also find this very appalling, to the point that I'd veto this. This rather points out that chrony itself should grow a standard directory where the admins can drop custom sources, instead of inventing a new directory here. This code does not even check if there is an existing `sources` in the file, i.e. whether adding our new one is even effective.

Also, we should really try hard to have a cockpit specific configuration file/dir -- our goal is to not be layered product on top of the OS, but just use *the* official OS API. If we have to touch chrony.conf anyway, then let's just read and write the custom NTP `server`s straight from there; or use said `-p` option to parse it (including sourced files). If chrony grows a default /etc/chrony/sources/ directory, we can of course use that one (similar to our timesyncd 50-cockpit.conf).

TBH, all of this is why https://bugzilla.redhat.com/show_bug.cgi?id=1330465 has been blocked for many years -- chrony's config format/structure is hard to machine-read/write. If we really still want to do this, then we need to do it properly IMHO, or first work with chrony to fix its config. But given how far you already got, looking for `server` entries does not actually sound that hard any more?

Reply:
> This code does not even check if there is an existing sources in the file, i.e. whether adding our new one is even effective.

Do you mean `sourcedir`?  There can be multiple `sourcedir` directives.

Reply:
> Also, we should really try hard to have a cockpit specific configuration file/dir -- our goal is to not be layered product on top of the OS, but just use the official OS API.

We use the identical approach with timesyncd: Drop our servers into a configuration snippet that we assume full control over.

I agree that the UI does not fully reflect this. Instead of saying "Automatically using specific NTP servers" we could say "Automatically using NTP with additional servers" to reflect that the default servers are still active.

Reply:
> But given how far you already got, looking for server entries does not actually sound that hard any more?

We also need to handle `pool` then and all the myriad options that can be added to those lines.

I would rather drop the whole "using specific NTP servers" option from our UI.

Reply:
I indeed meant `server` -- most admins would probably just edit chrony.conf itself to add them. For ansible etc. they might add a `sourcedir` indeed, with a bunch of files which may contain `server`  entries. That's what I meant with "we need to parse everything".

> We use the identical approach with timesyncd

It's a bit easier there: we don't have to define the equivalent of a `sourcedir`, as systemd provides a standard directory.

> reflect that the default servers are still active.

Hmm indeed -- if we do that for timesyncd at the moment, that sucks. It would e.g. not work for our own RH data center. We have to use *only* the specified NTP servers there, the default ones are unreachable. It seems that this is the much more common case?

Reply:
> It would e.g. not work for our own RH data center. We have to use only the specified NTP servers there, the default ones are unreachable. It seems that this is the much more common case?

But aren't unreachable servers just ignored?

Reply:
Certainly if they are firewalled, they "just" cause log messages. It still feels unexpected to query other ones if they get explicitly set in the UI. This works for timesyncd as its default `NTP=` is empty, so if a drop-in sets them, these are *the* ones that get used.

Reply:
> This works for timesyncd as its default NTP= is empty, so if a drop-in sets them, these are the ones that get used.

Unless there are more drop-in files or servers contributed by networkd.

Reply:
Right, it also ought to parse all drop-ins, and only write 50-cockpit.conf. dynamic NTP servers via DHCP kind of feel special -- I'm not actually sure whether or not one would want them, I can see both cases -- but I tend towards "use them", as DHCP and internal NTP servers are both under the local admin control.

Reply:
I feel this is a whole 'nother kettle of fish, improving what we do for timesyncd...

Reply:
Yes, I think we should fix the timesync parsing  to include all drop-ins, but that can/should happen in a different PR. I'm just saying that we shouldn't repeat the same error with chrony -- especially as it's a lot more muddy there.